### DISCUSSION TITLE
Axum and tracing to a file

### DISCUSSION THREAD

**Discussion by kellytk** [tokio-rs/tracing#1916] [Q&A]
[referred from [https://github.com/tokio-rs/axum/discussions/752#discussioncomment-2142874](https://github.com/tokio-rs/axum/discussions/752#discussioncomment-2142874)]

I have an [Axum](https://github.com/tokio-rs/axum)-based project successfully tracing to a file with one exception, request tracing.

I've followed the [tracing-aka-logging](https://github.com/tokio-rs/axum/tree/main/examples/tracing-aka-logging) example to make https://github.com/kellytk/axum-trace. The code runs as expected tracing to the terminal, however if I comment out [this line](https://github.com/kellytk/axum-trace/blob/main/src/main.rs#L6) and uncomment the following line, request handling traces do not appear in the file.

How please should it be correctly configured?

**Comment by hawkw** [comment]
One thing I noticed is that in your code for configuring tracing, you are configuring your `fmt` subscriber to filter spans and events based on a max level (`DEBUG`): https://github.com/kellytk/axum-trace/blob/b3f4018fa3a44bf0c49c24ba78fbed3df217570b/src/main.rs#L33. However, when you just call `tracing_subscriber::fmt::init`, it defaults to using an [`EnvFilter`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html) [configured from the `RUST_LOG` environment variable](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/fn.init.html). I notice that you are setting a value for this environment variable [here](https://github.com/kellytk/axum-trace/blob/b3f4018fa3a44bf0c49c24ba78fbed3df217570b/src/main.rs#L3-L5) if it is unset. However, in your `tracer()` function, that environment variable will not be used.

Try changing the call to `with_max_level(...)` to 
```rust
// ...
   .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
   // ...
```

You may need to enable the "env-filter" feature flag of the `tracing-subscriber` crate, if it is not already enabled.

**Reply by kellytk** [reply]
Thank you @hawkw.  I've made the change you suggested with no difference in behavior.  The "listening on..." event is traced to file however request traces are not.  I have confirmed the [presence](https://github.com/kellytk/axum-trace/blob/main/Cargo.toml#L11-L13) of the `env-filter` feature flag.  If you have further ideas I'd appreciate it.

The level overlap is from mixing the example's code into my own.

**Reply by hawkw** [reply]
Which specific traces do you see when using `fmt::init()` but not when logging to the file? Can you share the output in both cases or something?

**Reply by kellytk** [reply]
@hawkw I've added the output of each at https://github.com/kellytk/axum-trace.  I've also pushed your prior [suggestion](https://github.com/kellytk/axum-trace/blob/main/src/main.rs#L34).
